File: taxmate/uae_corporate_tax/report/uae_corporate_tax_worksheet/uae_corporate_tax_worksheet.js
// Copyright (c) 2026, Sourav Singh and contributors
frappe.query_reports["UAE Corporate Tax Worksheet"] = {
	filters: [
		{
			fieldname: "company",
			label: __("Company"),
			fieldtype: "Link",
			options: "Company",
			reqd: 1,
			default: frappe.defaults.get_user_default("Company"),
			on_change() {
				const company = frappe.query_report.get_filter_value("company");
				if (!company) {
					return;
				}
				frappe.call({
					method: "taxmate.uae_corporate_tax.utils.corporate_tax.get_ct_elections",
					args: { company },
					callback(r) {
						if (!r.message) {
							return;
						}
						frappe.query_report.set_filter_value("elect_sbr", r.message.elect_sbr);
						frappe.query_report.set_filter_value("elect_qfzp", r.message.elect_qfzp);
					},
				});
			},
		},
		{
			fieldname: "from_date",
			label: __("Period Start"),
			fieldtype: "Date",
			reqd: 1,
			default: frappe.datetime.year_start(),
		},
		{
			fieldname: "to_date",
			label: __("Period End"),
			fieldtype: "Date",
			reqd: 1,
			default: frappe.datetime.year_end(),
		},
		{
			fieldname: "elect_sbr",
			label: __("Elect Small Business Relief"),
			fieldtype: "Check",
			default: 0,
		},
		{
			fieldname: "elect_qfzp",
			label: __("Elect QFZP"),
			fieldtype: "Check",
			default: 0,
		},
	],
};

File: taxmate/uae_corporate_tax/report/uae_corporate_tax_worksheet/uae_corporate_tax_worksheet.py
# Copyright (c) 2026, Sourav Singh and contributors
# For license information, please see license.txt

"""Live Corporate Tax worksheet from GL + elections (does not file)."""

from __future__ import annotations

import frappe
from frappe import _
from frappe.utils import getdate


def execute(filters=None):
	if not frappe.has_permission("UAE CT Filing Log", "read"):
		frappe.throw(_("Not permitted"), frappe.PermissionError)

	filters = filters or {}
	columns = get_columns()
	if not filters.get("company") or not filters.get("from_date") or not filters.get("to_date"):
		return columns, []
	if getdate(filters["from_date"]) > getdate(filters["to_date"]):
		frappe.throw(_("From Date cannot be after To Date."))
	if not frappe.has_permission("Company", "read", filters["company"]):
		frappe.throw(_("Not permitted"), frappe.PermissionError)

	from taxmate.uae_corporate_tax.utils.corporate_tax import compute_ct

	result = compute_ct(
		filters["company"],
		filters["from_date"],
		filters["to_date"],
		elect_sbr=int(filters.get("elect_sbr") or 0),
		elect_qfzp=int(filters.get("elect_qfzp") or 0),
	)
	rows = [
		("Regime", result["regime"], ""),
		("Revenue", result["revenue"], _("SBR tests this, not profit")),
		("Expenses", result["expenses"], ""),
		("Accounting profit", result["accounting_profit"], _("From posted GL")),
		("Taxable profit", result["taxable_profit"], _("After add-backs / deductions on a Filing Log")),
		("Tax payable", result["tax_payable"], _("AED")),
		("Qualifying revenue", result["qualifying_revenue"], _("QFZP")),
		("Non-qualifying revenue", result["non_qualifying_revenue"], _("Includes unclassified")),
		("De minimis limit", result["de_minimis_limit"], _("Lower of 5% and AED 5m")),
		("De minimis passed", _("Yes") if result["de_minimis_passed"] else _("No"), ""),
		("Filing due", result["filing_due_date"], _("Period end + 9 months")),
	]
	data = [{"line": label, "amount": amount, "note": note} for label, amount, note in rows]
	return columns, data


def get_columns():
	return [
		{"fieldname": "line", "label": _("Line"), "fieldtype": "Data", "width": 220},
		{"fieldname": "amount", "label": _("Value"), "fieldtype": "Data", "width": 180},
		{"fieldname": "note", "label": _("Note"), "fieldtype": "Data", "width": 320},
	]
